rlpx/rlpx.go

```diff
@@ -0,0 +1,439 @@
+// Implements the RLPx protocol. Code is based off the devp2p spec defined here:
+// https://github.com/ethereum/devp2p/blob/master/rlpx.md
+package rlpx
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"hash"
+	mrand "math/rand"
+	"net"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/indexsupply/x/bint"
+	"github.com/indexsupply/x/ecies"
+	"github.com/indexsupply/x/enr"
+	"github.com/indexsupply/x/isxerrors"
+	"github.com/indexsupply/x/isxhash"
+	"github.com/indexsupply/x/isxsecp256k1"
+	"github.com/indexsupply/x/rlp"
+)
+
+type session struct {
+	Verbose bool
+
+	conn   net.Conn
+	local  *enr.Record
+	ig, eg *mstate
+}
+
+func (s *session) log(format string, args ...any) {
+	if s.Verbose {
+		fmt.Printf(format, args...)
+	}
+}
+
+func Session(l *enr.Record, hs *Handshake) (*session, error) {
+	var (
+		s   = &session{local: l}
+		err error
+	)
+
+	//static-shared-secret = ecdh.agree(privkey, remote-pubk)
+	//ephemeral-key = ecdh.agree(ephemeral-privkey, remote-ephemeral-pubk)
+	//shared-secret = keccak256(ephemeral-key || keccak256(nonce || initiator-nonce))
+	//aes-secret = keccak256(ephemeral-key || shared-secret)
+	//mac-secret = keccak256(ephemeral-key || aes-secret)
+	ephKey := secp256k1.GenerateSharedSecret(
+		hs.localEphPrvKey,
+		hs.remoteEphPubKey,
+	)
+	sharedSecret := isxhash.Keccak(append(
+		ephKey,
+		isxhash.Keccak(append(
+			hs.recipientNonce[:],
+			hs.initNonce[:]...,
+		))...,
+	))
+	aesSecret := isxhash.Keccak(append(ephKey, sharedSecret...))
+	macSecret := isxhash.Keccak(append(ephKey, aesSecret...))
+
+	s.ig, err = newmstate(aesSecret, macSecret)
+	if err != nil {
+		return nil, err
+	}
+	s.eg, err = newmstate(aesSecret, macSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	var inonce, rnonce [16]byte
+	for i := 0; i < 16; i++ {
+		inonce[i] = macSecret[i] ^ hs.initNonce[i]
+		rnonce[i] = macSecret[i] ^ hs.recipientNonce[i]
+	}
+
+	if hs.initiator {
+		//egress-mac = keccak256.init((mac-secret ^ recipient-nonce) || auth)
+		//ingress-mac = keccak256.init((mac-secret ^ initiator-nonce) || ack)
+		s.eg.hash = sha3.NewLegacyKeccak256()
```

Comment:
Isn't the `eg.hash` always sha3.NewLegacyKeccak256 by construction of the mstate?